backend/src/apiserver/server/util.go

```diff
@@ -216,9 +216,25 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
 	return nil
 }
 
+func validatePipelineSpecAndCheckPipelineVersionReference(resourceManager *resource.ResourceManager, spec *api.PipelineSpec, resourceReferences []*api.ResourceReference) error {
+	pipelineVersionId, err := CheckPipelineVersionReference(resourceManager, resourceReferences)
+	if err != nil {
+		return err
+	}
+	if (spec == nil || (spec.GetPipelineId() == "" && spec.GetWorkflowManifest() == "")) && pipelineVersionId == "" {
+		return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline spec or/and pipeline version.")
+	}
+	if spec != nil && spec.GetWorkflowManifest() != "" && pipelineVersionId != "" {
+		return util.NewInvalidInputError("Please either specify a pipeline version or a workflow manifest, not both.")
+	}
+	if err := ValidatePipelineSpec(resourceManager, spec); err != nil {
+		return err
+	}
+	return nil
+}
 func ValidatePipelineSpec(resourceManager *resource.ResourceManager, spec *api.PipelineSpec) error {
```

Comment:
Just wondering why `ValidatePipelineSpec` is public. Looks like it is only used in this package.

Reply:
good catch! You can change methods back to private when you see fit